File: packages/engine-community/src/detectors/RootDetector.ts
import type { Detector, DetectorResult } from './Detector.js';

const ROOT_BINARIES = ['su', 'busybox', 'magisk', 'daemonsu'] as const;
const ROOT_PATHS = [
  '/system/app/Superuser.apk',
  '/system/xbin/su',
  '/system/bin/su',
  '/sbin/su',
  '/data/local/xbin/su',
  '/data/local/bin/su',
  '/data/local/su',
] as const;

export class RootDetector implements Detector {
  readonly threatId = 'root' as const;
  readonly severity = 'high' as const;

  async run(): Promise<DetectorResult> {
    // In a real RN runtime this delegates to native via TurboModule.
    // In Node test environment we simulate based on environment variables
    // so tests can exercise the scoring logic without native code.
    const simulateRoot = process.env['GUARDIAN_SIMULATE_ROOT'] === '1';
    if (!simulateRoot) {
      return { detected: false, confidence: 0.1, evidence: { method: 'js-stub' } };
    }
    return {
      detected: true,
      confidence: 0.95,
      evidence: {
        method: 'simulated',
        paths: ROOT_PATHS.join(','),
        binaries: ROOT_BINARIES.join(','),
      },
    };
  }
}

File: packages/engine-community/android/src/main/kotlin/com/guardian/engine/community/RootDetector.kt
package com.guardian.engine.community

import java.io.File

object RootDetector {

    private val ROOT_PATHS = listOf(
        "/system/app/Superuser.apk",
        "/system/xbin/su",
        "/system/bin/su",
        "/sbin/su",
        "/data/local/xbin/su",
        "/data/local/bin/su",
        "/data/local/su",
        "/system/bin/.ext/.su",
        "/system/usr/we-need-root/su-backup",
    )

    private val ROOT_BINARIES = listOf("su", "busybox", "magisk", "daemonsu")

    data class Result(
        val detected: Boolean,
        val confidence: Float,
        val evidence: Map<String, String>,
    )

    fun detect(): Result {
        val foundPaths = ROOT_PATHS.filter { File(it).exists() }
        val foundBinaries = ROOT_BINARIES.filter { binary ->
            System.getenv("PATH")
                ?.split(":")
                ?.any { dir -> File(dir, binary).exists() } == true
        }

        val detected = foundPaths.isNotEmpty() || foundBinaries.isNotEmpty()
        val confidence = when {
            foundPaths.size >= 2 -> 0.97f
            foundPaths.size == 1 -> 0.80f
            foundBinaries.isNotEmpty() -> 0.70f
            else -> 0.05f
        }

        return Result(
            detected = detected,
            confidence = confidence,
            evidence = buildMap {
                if (foundPaths.isNotEmpty()) put("paths", foundPaths.joinToString(","))
                if (foundBinaries.isNotEmpty()) put("binaries", foundBinaries.joinToString(","))
            },
        )
    }
}
